rclpy/rclpy/impl/rcutils_logger.py

```diff
@@ -0,0 +1,286 @@
+        'throttle_time_source_type': 'RCUTILS_STEADY_TIME',
+    }
+
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        super(Throttle, cls).initialize_context(context, **kwargs)
+        context['throttle_last_logged'] = 0
+        # TODO(dhood): clarify the accepted values of the time source type, once it's actually used
+        if context['throttle_time_source_type'] != 'RCUTILS_STEADY_TIME':
+            raise ValueError(
+                'Received throttle_time_source_type of "{0}"; '
+                'currently only RCUTILS_STEADY_TIME is supported'
+                # TODO(dhood): remove the following note when the time source is actually rcutils
+                ' (simulated with python time.monotonic)'
+                .format(context['throttle_time_source_type']))
+
+    @staticmethod
+    def should_log(context):
+        logging_condition = True
+        # TODO(dhood): use rcutils time once support is available in rclpy
+        now = time.monotonic()
+        next_log_time = context['throttle_last_logged'] + context['throttle_duration_sec']
+        logging_condition = now >= next_log_time
+        if logging_condition:
+            context['throttle_last_logged'] = now
+        return logging_condition
+
+
+class SkipFirst(LoggingFilter):
+    """Ignore the first log call but process all subsequent calls."""
+
+    params = {
+        'skip_first': None,
+    }
+
+    @classmethod
+    def initialize_context(cls, context, **kwargs):
+        super(SkipFirst, cls).initialize_context(context, **kwargs)
+        context['first_has_been_skipped'] = False
+
+    @staticmethod
+    def should_log(context):
+        logging_condition = True
+        if not context['first_has_been_skipped']:
+            logging_condition = False
+            context['first_has_been_skipped'] = True
+        return logging_condition
+
+
+# The ordering of this dictionary defines the order in which filters will be processed.
+supported_filters = OrderedDict()
+supported_filters['throttle'] = Throttle
+supported_filters['skip_first'] = SkipFirst
+supported_filters['once'] = Once
+
+
+def get_filters_from_kwargs(**kwargs):
+    """
+    Determine which filters have had parameters specified in the given keyword arguments.
+
+    Returns the list of filters using the order specified by `supported_filters`.
+    """
+    detected_filters = []
+    all_supported_params = []
+    for supported_filter, filter_class in supported_filters.items():
+        filter_params = filter_class.params.keys()
+        all_supported_params.extend(filter_params)
+        if any(kwargs.get(param_name) for param_name in filter_params):
+            detected_filters.append(supported_filter)
+    # Check that all required parameters (with no default value) have been specified
+    for detected_filter in detected_filters:
+        for param_name, default_value in supported_filters[detected_filter].params.items():
+            if param_name not in kwargs:
```

Comment:
To avoid deep nesting I would suggest:

```
if param_name in kwargs:
    continue
```

Reply:
nesting avoided in 2c2e9208b7c7dc26382c1a8c00e1e86730cfde8e